asteroid/losses/pit_wrapper.py

```diff
@@ -191,6 +221,169 @@ def best_perm_from_perm_avg_loss(loss_func, est_targets, targets, **kwargs):
+        assert len(min_loss_indexes) == est_mixes.shape[0]
+
+        # For each batch there is a different min_loss_idx
+        ordered = torch.zeros_like(est_mixes)
+        for b, idx in enumerate(min_loss_indexes):
+            right_partition = parts[idx]
+            # sum the estimated sources to get the estimated mixtures
+            ordered[b, :, :] = torch.stack([torch.sum(est_targets[b, indexes, :][None, :, :], axis=1) for indexes in right_partition], axis=1)
+
+        return min_loss, ordered
+
+    @staticmethod
+    def best_part_mix_it_generalized(loss_func, est_targets, targets, **kwargs):
+        """ Find best partition of the estimated sources that gives
+            the minimum loss for the MixIT training paradigm in [1].
+            Valid only for two mixtures, but those mixtures do not
+            necessarly have to contain the same number of sources.
+            It is allowed the case where one mixture is silent.
+
+        Args:
+            loss_func: function with signature (targets, est_targets, **kwargs)
+                The loss function batch losses from.
+            est_targets: torch.Tensor. Expected shape [batch, nsrc, *].
+                The batch of target estimates.
+            targets: torch.Tensor. Expected shape [batch, nsrc, *].
+                The batch of training targets.
+            **kwargs: additional keyword argument that will be passed to the
+                loss function.
+
+        Returns:
+            tuple:
+                :class:`torch.Tensor`: The loss corresponding to the best
+                permutation of size (batch,).
+
+                :class:`torch.LongTensor`: The indexes of the best permutations.
+
+        References:
+            [1] Scott Wisdom and Efthymios Tzinis and Hakan Erdogan and Ron J Weiss
+            and Kevin Wilson and John R Hershey, "Unsupervised sound separation using
+            mixtures of mixtures." arXiv preprint arXiv:2006.12701 (2020) $
+        """
+
+        # check input dimensions
+        assert est_targets.shape[0] == targets.shape[0]
+        assert est_targets.shape[2] == targets.shape[2]
+
+        # get L, N and M
+        L = targets.shape[1]        # number of mixtures 
+        N = est_targets.shape[1]    # number of estimated sources
+
+        if L != 2:
+            raise ValueError('Works only with two mixtures')
+
+        # Generate all unique partitions of any size from a list lst of
+        # length n. Algorithm recursively distributes items over parts
+        def all_combinations(lst):
+            all_combinations = []
+            for k in range(len(lst) + 1):
+                for c in combinations(lst, k):
+                    rest = [x for x in lst if x not in c]
+                    all_combinations.append([list(c), rest]) 
+            return all_combinations
+
+        # Generate all the possible partitions
+        parts = all_combinations(range(N))    
+        for p, partition in enumerate(parts):
```

Comment:
I guess this could be factored out between the two methods?

Reply:
not really because the two functions are different...

Reply:
If you run this little example the difference will be clearer I guess:

```
# Exmple
n_mix = 2
n_est = 4
n_src = 4 // 2

# Mixit
def combs(lst, k, l):
    if l == 0:
        yield []
    else:
        for c in combinations(lst, k):
            rest = [x for x in lst if x not in c]
            for r in combs(rest, k, l-1):
                yield [list(c), *r]

parts = list(combs(range(n_est), n_src, n_mix))    
print('Mixit') 
for partition in parts:
    print(partition)

# Mixit_gen
def all_combinations(lst):
    all_combinations = []
    for k in range(len(lst) + 1):
        for c in combinations(lst, k):
            rest = [x for x in lst if x not in c]
            all_combinations.append([list(c), rest]) 
    return all_combinations

parts = all_combinations(range(n_est))  
print('Mixit_gen')  
for partition in parts:
    print(partition)
```

Reply:
Thanks for the example. 
I meant, the code under it. 
Yes, the mixing matrices/partitions are different, but the loop seems quite similar right?

Reply:
yes then the loop it's the same!

Reply:
also the number of sources and number of mixtures are pretty much always known beforehand, maybe we can cache the mixing matrix at beginning.

Reply:
Yes, caching the mixing matrix will be useful!